### DISCUSSION TITLE
Change "power_demand_divider" and "buffer" value without reflashing

### DISCUSSION THREAD

**Discussion by pfeupfeu** [syssi/esphome-soyosource-gtn-virtual-meter#111] [Q&A]
Since yersterday my setup is working! Now I would like to implement small improvements.

In winter, with little solar yield, only one of my 3 Soyos should run to get a better efficency. 
In summer, when there is "too much" solar yield, I would like to overfeed into the grid.

Therefore, I would like to be able to change the "power_demand_divider" and the "buffer" value without reflashing. 
Does anyone have a tip on how I can achieve this?

**Comment by syssi** [comment]
I will have a look how to make there settings templatable. At the moment individual values while runtime aren't supported because the values are assigned at compile time. Did I understand your setup correctly: You have attached three inverters sharing the same RS485 bus to one ESP and you turn off two of them at the winter?

**Comment by pfeupfeu** [comment]
Yes, I have attached three inverters sharing the same RS485 bus to one ESP and I turn off two of them at the winter. 
The Soyos are supposed to be switched off via a smart socket, for example.
The aim is to keep the standby consumption low and the efficiency high. The Soyos seem to have their max. efficiency at about 430W.

**Reply by Chickenbreast0** [reply]
> Yes, I have attached three inverters sharing the same RS485 bus to one ESP and I turn off two of them at the winter. The Soyos are supposed to be switched off via a smart socket, for example. The aim is to keep the standby consumption low and the efficiency high. The Soyos seem to have their max. efficiency at about 430W.


@pfeupfeu  syssi published an updated *yaml with buffer function. Maybe you saw it already ;)

I would realize the switching off of two inverters with an automation in Homeassistant: For example with a calendar trigger, which switches off the devices according to the season.

In my case, since I have only one panel with over 500W peak, I use automations to control the max power demand - depending on the solar power and as dependency the battery voltage, which I read directly from the charge controller (EPEVER). The voltage of the battery display in the Soyo and in the EPEVER has a voltage delta of 0.5V. Therefore, the value from the EPEVER is much more accurate.

**Reply by pfeupfeu** [reply]
I have adjusted the voltage measurement in the Soyo so that the Soyo's own measurement is now good enough:

The Soyo uses its 5V operating voltage, coming  from a 78M05, as its reference voltage. 
I replaced the 78M05 with a small ajustable DC/DC converter and set the voltage to 4.9V - then the DC voltage measurement is correct and additionaly the standby consumption has dropped by 1 Watt.

I don't have home assistant yet, so I try to do as much as possible with the ESP & Soyo.

**Comment by pfeupfeu** [comment]
Ahead: My programming skills are very limited. 
I like  to be fed more power into the grid when the battery is full. 
To do this, I want the ESP to decrease the `buffer ` by e.g. 100 when the battery voltage is high, and increase it again to the original `buffer ` value when the battery voltage drops.

I have written the following for this purpose:

```
globals:
   - id: bat_full
     type: bool
     restore_value: no
     initial_value: 'false'

sensor:
  - platform: template
    name: "Battery Full"
    id: batteryfull
    lambda: |-
        if ((id(bat_full) == 0) && id(battery_voltage0).state > 54.6)  {
          // Batterie ist jetzt voll, buffer herabsetzen
          id(bat_full) = 1;
          id(buffer0).state -= 100;
          return id(bat_full);
        } 
        if (id(battery_voltage0).state < 53.3 && id(bat_full) == 1) {
          // Batterie ist nicht mehr voll, buffer anheben
          id(bat_full) = 0;
          id(buffer0).state += 100;
          return id(bat_full);
        }
        return id(bat_full);
    update_interval: 10s

number:
  - platform: soyosource_virtual_meter
    soyosource_virtual_meter_id: virtualmeter0
    buffer:
      name: "${name} buffer W"
      id: buffer0
```

`buffer` is adjusted as desired in the web interface @ hight battery voltage, but the new buffer value is not used for the power_demand calculation. The calculation does not change. 
What I'm doing wrong?

**Reply by syssi** [reply]
I will have a look tomorrow. May be the buffer value isn't respected properly at the moment.

**Reply by syssi** [reply]
@pfeupfeu Which version do you use at the moment? The official one or the feature branch of @kev300?

**Reply by syssi** [reply]
I assume applying a value to the `id(buffer0).state` could be the issue. Could you try `id(buffer0).set_value()` instead?

**Reply by pfeupfeu** [reply]
I'm using the official version.
 
With `id(buffer0).set_value()` I get this error:
`error: 'class esphome::soyosource_virtual_meter::SoyosourceNumber' has no member named 'set_value'; did you mean 'set_name'?`

Additional info:
If I change the `buffer` value by hand within the Webinterface , the change has the desired effect, the calculation uses the new value.

**Reply by syssi** [reply]
Next try:

```
auto call = id(my_number).make_call();
call.set_value(42);
call.perform();
```

**Reply by pfeupfeu** [reply]
Yes! This is working! 
It is not possible to increment/decrement the value directly within `call.set_value`, but I can calculate the desired value in advance and set it.
